workstreams/Runtime_Init_Finalize_EntryPoint_Workstream.md

```diff
@@ -0,0 +1,102 @@
+the beginning of and end of these quantum function calls invoked from host
+call-sites. The utility of this type of function-level functionality in the
+support of near-term remote job submission and flushing internal quantum
+instruction queues, as well as general callbacks for library runtime
+event systems.
+
+This workstream will define how QIR generators should mark a function as
+EntryPoint. QIR generators can mark exactly one function as `EntryPoint`
+via an LLVM attribute. The function can be described with any valid function
+name. If there aren't any functions marked as `EntryPoint` then a `main`
+function must be present to enable execution. For custom `EntryPoint`
+functions an appropriate LLVM Pass must be applied to introduce a `__start`
+or `__main` function that calls the original entry point function.
+
+Use cases may exist whereby QIR code is generated with a custom EntryPoint and
+needs to be executed with custom command line arguments that should serve as
+input to that EntryPoint function. The specification needs to define some
+specified way to map custom command line arguments of a general type to the
+input arguments of the function, as well as the introduction of a `main`
+function for the underlying C runtime to pick up for execution.
+
+## Dependencies & Related Projects
+
+A related workstream is to specify a base profile of the full specification.
+This work should ensure that future base profile work is also compliant with
+entrypoint definitions and runtime input, finalization, and parameterization.
+
+## Deliverable(s) & Expected Outcome
+
+The expected outcome of this workstream is an update to the existing QIR
+specification documentation that defines specific function declarations
+for program-level initialization and finalization, function-level
+initialization and finalization for quantum functions invoked from host
+call sites, and entry-point marker metadata and runtime parameter input.
+
+## Future Work (Out of Scope)
+
+This work will not define specific initialization or finalization tasks
+that a QIR implementation library should perform.
+
+## Working Group & Getting Involved
+
+Working group participants: TBD <br/>
+Working group chair: Alex McCaskey
+
+If you would like to contribute to the workstream, please contact
+[qiralliance@mail.com](mailto:qiralliance@mail.com).
+
+## Schedule
+
+Launch date: <br/>
+Estimated end date: <br/>
+Meeting schedule and/or channel(s) of communication:
+
+## Status & Discussions
+
+Current status: Workstream to be approved by steering committee
+
+The work and status will be tracked in the form of a [GitHub
+issue](https://github.com/qir-alliance/qir-spec/issues) on the [specification
+repository](https://github.com/qir-alliance/qir-spec). We encourage comments,
+inputs, and discussions on that issue.
+
+The GitHub issue with be tagged appropriately after approval by the steering
+committee.
+
+## Open Questions
```

Comment:
All very good questions. Here’s how i think of it: 

The goal of this workstream is to define one time initialization and finalization for executable QIR code. The term executable here may take on one of two meanings - QIR code lowered to object code and linked with a runtime library implementing the QIR specification (used often for QPU emulation), or QIR code used as some exchange format between host and control system and transpiled / lowered to executable code for the classical electronic drivers. For both I think that initialization and finalization are useful concepts. Initialization has to be considered at the start of the EntryPoint function. Of course, a QIR code may not define an EntryPoint, in which case all functions are simply library functions. There is no reason why this could not be further lowered to object code and archived into a static or shared library and linked with other QIR libraries and / or an executable QIR code (contains EntryPoint function). 

The workstream should specify how one marks a LLVM function as EntryPoint (or default to main if it is present). If this marker is present, then the QIR code can be considered as executable. In the case of runtime library linking, execution comes naturally via the system C runtime library. In the case of QIR submission / codegen lowering it should be on control system developers to define how execution happens, but they will be able to key on the EntryPoint marker to understand where programmers expect execution to begin. 

I think linking QIR + classical object code can occur naturally with standard linker tools (especially for QIR runtime library linking, i.e. in the case of simulation). For physical backends, more care needs to be taken (I expect this will be discussed in workstream meetings) with regards to what classical functionality can be present on control system hardware. 

Of course this feeds into the Base Profile workstream - I think that QIR init / finalize concepts can fit naturally there as well. In the case of physical system codegen, lowering tools can either ignore the init / finalize functions (treat them as markers) or they could pull in pertinent init / finalize tasks for the hardware as linked functions at runtime.